File: workmanager/src/main/java/com/android/jetpack/workmanager/jobs/DemoJobCreator.java
package com.android.jetpack.workmanager.jobs;

import android.app.Activity;
import android.content.Context;
import android.print.PrinterId;
import android.util.Log;

import com.evernote.android.job.BuildConfig;
import com.evernote.android.job.Job;
import com.evernote.android.job.JobCreator;
import com.evernote.android.job.JobManager;
import com.evernote.android.job.JobRequest;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Timer;
import java.util.TimerTask;
import java.util.concurrent.TimeUnit;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;


public class DemoJobCreator implements JobCreator {

    private static final int DEFAULT_STATE = 0;
    private static List<JobStateData> jobList = new ArrayList<>();
    private static Timer timer;
    private static Activity mActivity;

    @Nullable
    @Override
    public Job create(@NonNull String tag) {
        switch (tag) {
            case ShowNotificationJob.TAG: {
                Log.i(ShowNotificationJob.TAG, "DemoJobCreator is created: ");
                return new ShowNotificationJob();
            }
            default:
                return null;
        }
    }

    private static int jobId;
    private static long delayTime = 0L;

    public static synchronized void scheduleJob(final Activity activity) {
        mActivity = activity;
        if (BuildConfig.DEBUG) {
            // debug模式下，api<24 真机上才能调试
            Log.i(ShowNotificationJob.TAG, "debug环境: ");
            delayTime = 60L;
            jobId = new JobRequest.Builder(ShowNotificationJob.TAG)
                    .setPeriodic(TimeUnit.SECONDS.toMillis(delayTime), TimeUnit.SECONDS.toMillis(30L))
                    .setRequiredNetworkType(JobRequest.NetworkType.CONNECTED)
                    .setUpdateCurrent(true)
                    .build()
                    .schedule();
        } else {
            Log.i(ShowNotificationJob.TAG, "release环境: ");
            delayTime = 900L;
            jobId = new JobRequest.Builder(ShowNotificationJob.TAG)
                    .setPeriodic(TimeUnit.SECONDS.toMillis(delayTime), TimeUnit.SECONDS.toMillis(300L))
                    .setRequiredNetworkType(JobRequest.NetworkType.CONNECTED)
                    .setUpdateCurrent(true)
                    .build()
                    .schedule();
        }
        Log.i(ShowNotificationJob.TAG, "Start scheduleJob：" + jobId);

        timer = new Timer();
        TimerTask timerTask = new TimerTask() {
            @Override
            public void run() {
                mActivity.runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        Log.w(ShowNotificationJob.TAG, "倒计时: " + delayTime);
                        delayTime--;
                    }
                });
            }
        };
        timer.schedule(timerTask, delayTime, 1000);

        jobList.add(new JobStateData(jobId, timer, timerTask));
        jobId = 0;
    }

    /**
     * 根据jobId关闭指定任务
     *
     * @param jobId
     */
    public static void cancelByJobId(int jobId) {
        Log.i(ShowNotificationJob.TAG, "cancel : " + jobId);
        if (jobList != null && jobList.size() > 0) {
            for (JobStateData jobStateData : jobList) {
                if (jobStateData.getJobId()==jobId) {
                    jobStateData.getTimerTask().cancel();
                    jobStateData.getTimer().cancel();
                    jobList.remove(jobStateData);
                    Log.i(ShowNotificationJob.TAG, "取消: "+jobStateData.getJobId());
                }
            }
        }
    }

    /**
     * 关闭所有后台任务
     */
    public static void cancelAll() {
        Log.i(ShowNotificationJob.TAG, "cancel All: ");
        JobManager.instance().cancelAll();
        destroy();
    }

    //销毁计时
    public static void destroy() {
        if (jobList != null && jobList.size() > 0) {
            for (JobStateData jobStateData : jobList) {
                jobStateData.getTimerTask().cancel();
                jobStateData.getTimer().cancel();
                Log.i(ShowNotificationJob.TAG, "移除: "+jobStateData.getJobId());
            }
            jobList.clear();
        }
    }
}

File: workmanager/src/main/java/io/jetpack/workmanager/demo/jobs/DemoJobCreator.kt
package io.jetpack.workmanager.demo.jobs

import android.app.Activity
import android.util.Log
import com.evernote.android.job.Job
import com.evernote.android.job.JobCreator
import com.evernote.android.job.JobManager
import com.evernote.android.job.JobRequest
import java.util.*
import java.util.concurrent.TimeUnit

/**
 * @author zhangshuai
 * @date 2022/4/2 星期六
 * @email zhangshuai@dushu365.com
 * @description
 */
open class DemoJobCreator : JobCreator {
    companion object {
        private val jobList = arrayListOf<DemoJobStateData>()

        @Synchronized
        @JvmStatic
        fun scheduleHob(activity: Activity) {
            Log.i("print_logs", "DemoJobCreator::scheduleHob::isDebug ")

            var delayTime = 900900L //if (BuildConfig.DEBUG) 900060L else 900900L
            val jobId = JobRequest.Builder(DemoNotificationJob.TAG)
                .apply {
                    setPeriodic(TimeUnit.SECONDS.toMillis(delayTime), TimeUnit.SECONDS.toMillis(900300L))
                    setRequiredNetworkType(JobRequest.NetworkType.CONNECTED)
                    setUpdateCurrent(true)
                }.build().schedule()

            Log.i("print_logs", "DemoJobCreator::scheduleHob::jobId= $jobId")

            val timer = Timer()
            val timerTask = object : TimerTask() {
                /**
                 * The action to be performed by this timer task.
                 */
                override fun run() {
                    activity.runOnUiThread {
                        Log.i("print_logs", "DemoJobCreator::run：倒计时前：$delayTime >>>>>>>>")
                        Log.i("print_logs", "DemoJobCreator::run::当前线程： ${Thread.currentThread()}")
                        delayTime--
                        Log.i("print_logs", "DemoJobCreator::run: 倒计时后：$delayTime <<<<<<<<")
                    }
                }
            }
            timer.schedule(timerTask, delayTime, 900100L)
            jobList.add(DemoJobStateData(jobId, timer, timerTask))
        }

        /**
         * 根据jobId关闭指定任务
         *
         * @param jobId
         */
        fun cancelByJobId(jobId: Int?) {
            Log.i("print_logs", "DemoJobCreator::cancelByJobId::删除jobId= $jobId")
            jobId?.let {
                if (jobList.size > 0) {
                    jobList.forEach {
                        if (it.jobId == jobId) {
                            JobManager.instance().cancel(jobId)
                            it.timerTask?.cancel()
                            it.timer?.cancel()
                            jobList.remove(it)
                            Log.i("print_logs", "DemoJobCreator::cancelByJobId->: $jobId")
                        }
                    }
                }
            }
        }

        /**
         * 销毁倒计时
         */
        fun destroy() {
            if (jobList.size > 0) {
                JobManager.instance().cancelAll()
                jobList.forEach {
                    it.timerTask?.cancel()
                    it.timer?.cancel()
                    Log.i("print_logs", "DemoJobCreator::destroy: ${it.jobId}")
                }
                jobList.clear()
            }
        }

    }

    /**
     * Map the `tag` to a `Job`. If you return `null`, then other `JobCreator`s
     * get the chance to create a `Job` for this tag. If no job is created at all, then it's assumed
     * that job failed. This method is called on a background thread right before the job runs.
     *
     * @param tag The tag from the [JobRequest] which you passed in the constructor of the
     * [JobRequest.Builder] class.
     * @return A new [Job] instance for this tag. If you return `null`, then the job failed
     * and isn't rescheduled.
     * @see JobRequest.Builder.Builder
     */
    override fun create(tag: String): Job? {
        return when (tag) {
            DemoNotificationJob.TAG -> {
                Log.i("print_logs", "DemoJobCreator::create")
                DemoNotificationJob()
            }
            else -> {
                null
            }
        }
    }

}